dataframe merge returns incorrect result when data is from multi-partition parquet

Similar to issue #2237 , I am seeing dask dataframe merge result can be incorrect when data is from multi-partition parquet. pandas merge always produces correct results.

I have put my test data in https://github.com/ningsean/dask_merge_issue
I am running a local cluster at localhost:9898

My setup is Python 3.6.1, with the following module versions. 

```
dask.__version__
'0.15.2+25.ga1a75a7'

numpy.__version__
'1.13.1'

partd.__version__
'0.3.8'

pandas.__version__
'0.20.1'
```

Here is the code that shows the problem.
```
from dask.distributed import Client
import dask.dataframe as dd
import pandas as pd

client = Client("localhost:9898") 

ddf1 = dd.read_parquet("keys_133m.parq")
ddf1.columns
#Index(['KEY'], dtype='object')
ddf1.dtypes
#KEY    int64

ddf2 = dd.read_parquet("key_ids_133m.parq")
ddf2.columns
#Index(['KEY', 'ID'], dtype='object')
ddf2.dtypes
#KEY    int64
#ID     int64
```

### pandas inner merge result has about **500k** rows
```
df = ddf1.compute().merge(ddf2.compute(),how='inner',on='KEY')
df.shape
#(505847, 2)
```

### dask inner merge has only about **272k** rows
```
ddf = ddf1.merge(ddf2,how='inner',on='KEY')
len(ddf)
#272051
```

### One example of missing row in ddf is KEY==135337982:
```
#dask merge result does not contain the row
ddf.loc[ddf.KEY==135337982].compute()
#Empty DataFrame
#Columns: [KEY, ID]
#Index: []

#pandas merge result contains the row
df.loc[df.KEY==135337982]
#              KEY         ID
#122842  135337982  149107799
```

### Single-partition parquet works
```
#use nlargest to write to a single-partition parquet
ddf1.nlargest(n=505850,columns='KEY').to_parquet('keys_one_partition.parq')
ddf2.nlargest(n=1960255,columns='KEY').to_parquet('key_ids_one_partition.parq')

ddf1_one = dd.read_parquet('keys_one_partition.parq')
ddf2_one = dd.read_parquet('key_ids_one_partition.parq')

ddf_one = ddf1_one.merge(ddf2_one,how='inner',on='KEY')
len(ddf_one)
#505847

df_one = ddf1_one.compute().merge(ddf2_one.compute(),how='inner',on='KEY')
df_one.shape
#(505847, 2)
```

I haven't been able to reproduce this using synthetic data. However, the issue persists with the parquet files that @ningsean uploaded.
```python

import requests
import dask
import dask.dataframe as dd
import zipfile
r = requests.get("https://github.com/ningsean/dask_merge_issue/raw/master/data.zip" )
with open('data.zip', 'wb') as f:
    f.write(r.content)

zipfile.ZipFile('data.zip', 'r').extractall('.')

dd_right = dd.read_parquet('keys_133m.parq/', index=False)
dd_left = dd.read_parquet('key_ids_133m.parq/', columns = ['KEY', 'ID'], index=False)
with dask.set_options(shuffle='tasks'):
    dd_merged = dd_left.merge(dd_right, how='inner', on='KEY')
    df_computed = dd_merged.compute()

computed_from_pandas = dd_left.compute().merge(dd_right.compute(), how='inner', on='KEY')
assert len(df_computed) == len(computed_from_pandas)
out:
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-41-d28889968375> in <module>()
----> 1 assert len(df_computed) == len(computed_from_pandas)

AssertionError: 
```
Hope this is helpful.

This is a really weird one. It appears that the task shuffle breaks down when number of tasks increase to a big number. here is a reproducible example:
```python
In [2]: import numpy as np
   ...: import pandas as pd
   ...: import dask
   ...: import dask.dataframe as dd
   ...: large = pd.DataFrame({'KEY': np.arange(0,50000)})  
   ...: small = pd.DataFrame({'KEY': np.arange(2500,50000)})
   ...: dd_right = dd.from_pandas(large,npartitions=257)
   ...: dd_left = dd.from_pandas(small, npartitions=3)
   ...: with dask.set_options(shuffle='tasks'):
   ...:     dd_merged = dd_left.merge(dd_right, how='inner', on='KEY')
   ...:     result = dd_merged.compute()
   ...:     
   ...: expected = large.merge(small, how='inner', on='KEY')
   ...: 

In [3]: pd.testing.assert_frame_equal(result.sort_values('KEY').reset_index(drop=True), expected)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-3-e4e28f598970> in <module>()
----> 1 pd.testing.assert_frame_equal(result.sort_values('KEY').reset_index(drop=True), expected)

~/miniconda3/envs/dask_join_dev/lib/python3.6/site-packages/pandas/util/testing.py in assert_frame_equal(left, right, check_dtype, check_index_type, check_column_type, check_frame_type, check_less_precise, check_names, by_blocks, check_exact, check_datetimelike_compat, check_categorical, check_like, obj)
   1351                             'DataFrame shape mismatch',
   1352                             '{shape!r}'.format(shape=left.shape),
-> 1353                             '{shape!r}'.format(shape=right.shape))
   1354 
   1355     if check_like:

~/miniconda3/envs/dask_join_dev/lib/python3.6/site-packages/pandas/util/testing.py in raise_assert_detail(obj, message, left, right, diff)
   1091         msg += "\n[diff]: {diff}".format(diff=diff)
   1092 
-> 1093     raise AssertionError(msg)
   1094 
   1095 

AssertionError: DataFrame are different

DataFrame shape mismatch
[left]:  (47322, 1)
[right]: (47500, 1)
```

however, the following works just fine:
```python
In [6]: 
   ...: dd_right = dd.from_pandas(large, npartitions=256)
   ...: dd_left = dd.from_pandas(small, npartitions=3)
   ...: with dask.set_options(shuffle='tasks'):
   ...:     dd_merged = dd_left.merge(dd_right, how='inner', on='KEY')
   ...:     result = dd_merged.compute()
   ...:     
   ...: expected = large.merge(small, how='inner', on='KEY')
   ...: 

In [7]: pd.testing.assert_frame_equal(result.sort_values('KEY').reset_index(drop=True), expected)

```

--- a/dask/dataframe/multi.py
+++ b/dask/dataframe/multi.py
@@ -283,11 +283,12 @@ def hash_join(lhs, left_on, rhs, right_on, how='inner',
                      how, npartitions, suffixes, shuffle, indicator)
     name = 'hash-join-' + token
 
-    dsk = dict(((name, i), (methods.merge, (lhs2._name, i), (rhs2._name, i),
-                            how, left_on, right_on,
-                            left_index, right_index, indicator,
-                            suffixes, lhs._meta, rhs._meta))
-               for i in range(npartitions))
+    dsk = {(name, i): (methods.merge,
+                       (lhs2._name, i), (rhs2._name, i),
+                       how, left_on, right_on,
+                       left_index, right_index, indicator,
+                       suffixes, lhs._meta, rhs._meta)
+           for i in range(npartitions)}
 
     divisions = [None] * (npartitions + 1)
     return new_dd_object(toolz.merge(lhs2.dask, rhs2.dask, dsk),

--- a/dask/dataframe/shuffle.py
+++ b/dask/dataframe/shuffle.py
@@ -460,8 +460,7 @@ def shuffle_group(df, col, stage, k, npartitions):
     npartitions, k, stage = [np.array(x, dtype=np.min_scalar_type(x))[()]
                              for x in [npartitions, k, stage]]
 
-    c = np.mod(c, npartitions).astype(typ)
-    c = np.mod(c, npartitions, out=c)
+    c = np.mod(c, npartitions).astype(typ, copy=False)
     c = np.floor_divide(c, k ** stage, out=c)
     c = np.mod(c, k, out=c)
 
@@ -470,9 +469,7 @@ def shuffle_group(df, col, stage, k, npartitions):
     locations = locations.cumsum()
     parts = [df2.iloc[a:b] for a, b in zip(locations[:-1], locations[1:])]
 
-    result = dict(zip(range(k), parts))
-
-    return result
+    return dict(zip(range(k), parts))
 
 
 def shuffle_group_3(df, col, npartitions, p):

--- a/dask/dataframe/shuffle.py
+++ b/dask/dataframe/shuffle.py
@@ -203,6 +203,7 @@ def shuffle(df, index, shuffle=None, npartitions=None, max_branch=32,
     """
     if not isinstance(index, _Frame):
         index = df._select_columns_or_index(index)
+
     partitions = index.map_partitions(partitioning_index,
                                       npartitions=npartitions or df.npartitions,
                                       meta=pd.Series([0]))
@@ -363,6 +364,7 @@ def rearrange_by_column_tasks(df, column, max_branch=32, npartitions=None):
     if npartitions is not None and npartitions != df.npartitions:
         parts = [i % df.npartitions for i in range(npartitions)]
         token = tokenize(df2, npartitions)
+
         dsk = {('repartition-group-' + token, i): (shuffle_group_2, k, column)
                for i, k in enumerate(df2.__dask_keys__())}
         for p in range(npartitions):
@@ -442,6 +444,11 @@ def shuffle_group_get(g_head, i):
 
 
 def shuffle_group(df, col, stage, k, npartitions):
+    """ Splits dataframe into groups
+
+    The group is determined by their final partition, and which stage we are in
+    in the shuffle
+    """
     if col == '_partitions':
         ind = df[col]
     else:
@@ -449,11 +456,11 @@ def shuffle_group(df, col, stage, k, npartitions):
 
     c = ind._values
     typ = np.min_scalar_type(npartitions * 2)
-    c = c.astype(typ)
 
     npartitions, k, stage = [np.array(x, dtype=np.min_scalar_type(x))[()]
                              for x in [npartitions, k, stage]]
 
+    c = np.mod(c, npartitions).astype(typ)
     c = np.mod(c, npartitions, out=c)
     c = np.floor_divide(c, k ** stage, out=c)
     c = np.mod(c, k, out=c)
@@ -463,7 +470,9 @@ def shuffle_group(df, col, stage, k, npartitions):
     locations = locations.cumsum()
     parts = [df2.iloc[a:b] for a, b in zip(locations[:-1], locations[1:])]
 
-    return dict(zip(range(k), parts))
+    result = dict(zip(range(k), parts))
+
+    return result
 
 
 def shuffle_group_3(df, col, npartitions, p):

--- a/dask/dataframe/tests/test_shuffle.py
+++ b/dask/dataframe/tests/test_shuffle.py
@@ -5,6 +5,7 @@ import pickle
 import numpy as np
 import string
 from copy import copy
+import pandas.util.testing as tm
 
 import dask
 import dask.dataframe as dd
@@ -710,3 +711,23 @@ def test_remove_nans():
             params = [none_val if x is None else conv(x) for x in inputs]
             expected = [conv(x) for x in expected]
             assert remove_nans(params) == expected
+
+
+@pytest.mark.slow
+def test_gh_2730():
+    large = pd.DataFrame({'KEY': np.arange(0, 50000)})
+    small = pd.DataFrame({'KEY': np.arange(25, 500)})
+
+    dd_left = dd.from_pandas(small, npartitions=3)
+    dd_right = dd.from_pandas(large, npartitions=257)
+
+    with dask.set_options(shuffle='tasks', get=dask.get):
+        dd_merged = dd_left.merge(dd_right, how='inner', on='KEY')
+        result = dd_merged.compute()
+
+    expected = large.merge(small, how='inner', on='KEY')
+
+    tm.assert_frame_equal(
+        result.sort_values('KEY').reset_index(drop=True),
+        expected
+    )

--- a/docs/source/changelog.rst
+++ b/docs/source/changelog.rst
@@ -2,6 +2,27 @@ Changelog
 =========
 
 
+0.XX.X / 2018-MM-DD
+-------------------
+
+Array
++++++
+
+
+DataFrame
++++++++++
+
+-  Fixed bug in shuffle due to aggressive truncation (:pr:`3201`) `Matthew Rocklin`_
+
+
+Bag
++++
+
+
+Core
+++++
+
+
 0.17.1 / 2018-02-22
 -------------------
 

